### PR TITLE
SONiC Port mirroring HLD

### PR DESCRIPTION
**Pull request by rupesh-k**
SONiC Port Mirroring HLD

### CODE REVIEW

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+## 1.1 Functional Requirements
+
+1. Port/Port-channel mirroring support
+     - Add support to mirror ingress traffic on port/port-channel to SPAN/ERPSAN mirror session.
+     - Add support to mirror egress traffic on port/port-channel to SPAN/ERSPAN mirror session.
+     - Add support to mirror both ingress/egress traffic on port/port-channel to SPAN/ERSPAN mirror session.
+
+2. Dynamic session management
+    - Allow multiple source to single destination.
+    - Each session supports mirroring from single port to single destination port.
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: do we allow one mirror session from multiple source port to single dest port?

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+
+## 1.2 Configuration and Management Requirements
+- Existing CLI 'config mirror_session add/remove'to be extended to include source port/portchannel.
+- Existing CLI 'config mirror_session add/remove' to be extended to include destination port/portchannel.
+- Existing CLI 'show mirror session' is extended to support all flavors of mirror sessions.
+
+
+## 1.3 Scalability Requirements
+- Up to max ASIC capable mirror sessions to be supported.
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: how do we know how many mirror session can be supported in the asic?

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+- Once max mirror sessions are created and user attempts to create new session, error will be logged in syslog.
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: it is not discussed in the design document how this is implemented. can you add a section to discuss how you are going to implement this?

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+## 3.2 DB Changes
+### 3.2.1 CONFIG DB
+
+Existing table PORT_MIRROR_TABLE is enhanced to accept new source and destination configuration options  in the configuration database. This table is filled by the management framework.
+
+#### CONFIG_PORT_MIRROR_TABLE
+
+    ;Configure SPAN/ERSPAN mirror session.
+    ;storm control type - broadcast / unknown-unicast / unknown-multicast
+    key       = PORT_MIRROR_TABLE:mirror_session_name ; mirror_session_name is
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: PORT_MIRROR_TABLE -> PORT_MIRROR

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+                                                      ; unique session
+                                                      ; identifier
+    ;field  = value
+    destination_port = PORT_TABLE:ifname    ; ifname must be unique across PORT TABLE.
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: can dest port be lag?

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    source_port = PORT_TABLE:ifname    ; ifname must be unique across PORT,INTF,LAG TABLES
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: what do you mean by INTF? INTF is layer 3 concept.

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    direction     = ingress or egress or both           ; Direction ingress or egress or both.
+
+    mirror_session_name = 1*255VCHAR
+
+### 3.2.2 APP_DB
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: what is plan to support this feature in virtual switch, like sflow.

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+   - Populates the mirror attribute SAI structures and pushes the entry to ASIC_DB.·
+
+## 3.4 SAI
+Mirror SAI interface APIs are already defined. More details about SAI API and attributes are described below SAI Spec @
+
+https://github.com/opencomputeproject/SAI/blob/master/inc/saimirror.h
+
+## 3.5 CLI
+### 3.5.1 Data Models
+Custom Yang model will be introduced for this feature.
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: can you add description for the yang model?

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+    ERSPAN Sessions
+    ---------------------------------------------------------------------------------------------------------
+      Name     Status      SRC IP    DST IP    GRE   DSCP    TTL  Queue Policer      SRC Port    Direction
+    everflow0  active    10.1.0.32  10.0.0.7    10    10      10
+    everflow1  active    10.1.0.33  10.0.0.8    10    10      10                     Ethernet4    both
+
+    SPAN Sessions
+    ---------------------------------------------------------------------------------------------------------
+      Name Status         DST Port         SRC Port Direction
+     sess1 active        Ethernet4        Ethernet0     rx
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: what is the definition of active status? what is the criteria. can you make it clear in the document?

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+###### Table 3: Scaling limits
+|Name                      |   Scaling value    |
+|--------------------------|--------------------|
+| Max mirror sessions      | silicon specific   |
+
+# 9 Unit Test
+
+## 9.1 CLI Test Cases
+
+    1. Configure ERSPAN mirror session and verify all parameters are updated properly in CONFIG_DB
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: please describe where the tests are going to be contributed to? which repo?

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+    config mirror_session add erspan <session-name> <src_ip> <dst_ip> <gre> <dscp>  [ttl] [queue]
+
+    #Configure Destination only span mirror session.
+    config mirror_session add span <session-name> <destination_ifName>
+
+    # Modify existing ERSPAN configuration to accept source port and direction
+    config mirror_session add erspan <session-name> <src_ip> <dst_ip> <gre> <dscp>  [ttl] [queue] [src_port] [rx/tx/both]
+
+    #Configure Port mirror span mirror session.
+    config mirror_session add span <session-name> <destination_ifName> <source_ifName> <rx/tx/both>
```
Comment (lguohan) [inline doc/SONiC_Port_Mirroring_HLD.md]: is policer going to supported in span session?

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 5 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:140]: No. Destination port cant be LAG. We are not supporting this now.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 11 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:189]: Yes. Policer is supported.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 10 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:292]: Some of the tests will be contributed to swss and others will be contributed to spytest. I will update in the doc.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 9 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:218]: Port mirror session will be active in below cases.
1. When destination port only session is created, then once the session is created from SAI. the session becomes active. These sessions can be used in ACL mirroring.
2. When mirroring is enabled on the source ports, then the session will become active.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 8 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:173]: Sure. I will add and update the doc

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 2 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:101]: The SAI Attribute SAI_SWITCH_ATTR_MAX_MIRROR_SESSION can be used to support the max mirror sessinons. This is for both ERSPAN and SPAN and we don't have any mechanism in orchagent to retrieve this now.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 1 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:78]: Yes. we allow one mirror session from multiple source port to single destination port.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 4 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:136]: Done. Updated the doc.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 3 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:102]: When the user exceeds max mirror sessions then SAI returns SAI_STATUS_INSUFFICIENT_RESOURCES, This is similar behaviour as what is currently present in ERSPAN also. The error is logged to syslog from SyncD. OrchAgent treats the error as fatal, similar to the existing ERPSAN code.

**Comment by rupesh-k** [comment]
Thanks Guohan for review comments.
I have updated the document with details.
I will be updating more on unittests in next commit.

Thanks

**Comment by rupesh-k** [comment]
Hi @lguohan ,

Can u please review the HLD and I have raised initial code PR along with pytest UT 
Following are code PR links.
https://github.com/Azure/sonic-swss/pull/1314
https://github.com/Azure/sonic-utilities/pull/936

Can u please help review .

Thanks
Rupesh Kumar

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 6 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md]: Doc is already updated. this is only port for destination port, port/LAG for src port.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 7 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:144]: We dont have plan for this now. If u mean to handle specific flow to mirror, then ACL mirroring can be used. Can u please clarify on sflow part here.

**Comment by ben-gale** [comment]
@lguohan, @xinliu-seattle  - can we merge this now please?

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 8 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md]: Updated doc with both SONiC yang and openconfig extension model also

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 2 above)
Comment (dev-aviz) [reply doc/SONiC_Port_Mirroring_HLD.md:98]: Do we have a number based on TD3 and TH series?
Also how many sessions can be active at the same time. Is there any limitation

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 5 above)
Comment (dev-aviz) [reply doc/SONiC_Port_Mirroring_HLD.md:138]: Can you update the Source-ip and destination-ip fields to the table

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+```
+    # Modify existing ERSPAN configuration as below.
+    config mirror_session add erspan <session-name> <src_ip> <dst_ip> <gre> <dscp>  [ttl] [queue] --policer <policer>
+
+    #Configure Destination only span mirror session.
+    config mirror_session add span <session-name> <destination_ifName>
+
+    # Modify existing ERSPAN configuration to accept source port and direction
+    config mirror_session add erspan <session-name> <src_ip> <dst_ip> <gre> <dscp>  [ttl] [queue] [src_port] [rx/tx/both] --policer <policer>
+
```
Comment (dev-aviz) [inline doc/SONiC_Port_Mirroring_HLD.md:318]: Why do we have dscp here? Is it for any prioritization of ERSPAN traffic mirrored across the devices?

File: doc/SONiC_Port_Mirroring_HLD.md [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+# 8 Scalability
+
+Max mirror sessions supported are silicon specific. 
+
+###### Table 3: Scaling limits
+|Name                      |   Scaling value    |
+|--------------------------|--------------------|
+| Max mirror sessions      | silicon specific   |
+
```
Comment (dev-aviz) [inline doc/SONiC_Port_Mirroring_HLD.md:419]: Do you have any scaling numbers?

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 12 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:318]: Yes ERSPAN tunnel header will have this dscp, This is already supported in current community code and this PR doesnt modify any behaviour of this field.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 13 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:419]: For Broadcom ASICs, max 4 sessions can be supported.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 5 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:138]: ERSPAN is already supported in community and we captured only new additions which are done as part of this PR.

File: doc/SONiC_Port_Mirroring_HLD.md (hunk 2 above)
Comment (rupesh-k) [reply doc/SONiC_Port_Mirroring_HLD.md:98]: For TD3 and TH series, max 4 sessions can be active at same time. single session can be shared across multiple source ports.

**Pull request merged**